### PR TITLE
ci: refactor axon start with short genesis workflow

### LINKED ISSUE #1322: chore: refactor workflows/axon-start-with-short-genesis.yml
## Failed CI
https://github.com/axonweb3/axon/actions/runs/5865811220/job/15903430440

### Related PR
- https://github.com/axonweb3/axon/pull/1323

## TODOs for workflows/axon-start-with-short-genesis.yml

- [x] don't have to use [self-hosted runner](https://github.com/axonweb3/axon/blob/e6c4f6770073d7dc95d3075dd1ebd25481db0d6e/.github/workflows/axon-start-with-short-genesis.yml#L9)
- [x] (optional) don't push the test image to Docker Hub in this workflow
- [x] (optional) use health-check method of an Axon node, not [`grep from logs`](https://github.com/axonweb3/axon/blob/e6c4f6770073d7dc95d3075dd1ebd25481db0d6e/.github/workflows/axon-start-with-short-genesis.yml#L41C55-L41C59)

### PR DESCRIPTION
**Pull request by Flouse**
## What this PR does / why we need it?

`cargo build` takes a lot of time
=> This PR cache the building result with commit_id in the cache-key, so we don't have to build Axon from zero in every GitHub action workflow.
=> With cargo cache, the `Build Axon` step only takes 90 seconds.
=> The workflows such as openzeppelin_tests will consume ~20 mins less.

### What is the impact of this PR?

- resolve https://github.com/axonweb3/axon/issues/1322
- No Breaking Change
- improve efficiency in project building and testing，and may report bugs effectively
- Perhaps in the near future, all tests could be triggered by default, not manually.

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [x] Cargo Clippy
- [ ] Coverage Test
- [x] E2E Tests
- [x] Code Format
- [x] Unit Tests
- [x] Web3 Compatible Tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [x] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`                     |
| *Coverage Test*                           | Get the unit test coverage report                                         |
| *E2E Test*                                | Run end-to-end test to check interfaces                                   |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`           |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

File: .github/workflows/axon-start-with-short-genesis.yml [hunk 1]
```diff
@@ -1,59 +1,121 @@
...
+        EOF
+      timeout-minutes: 1
 
+    - name: Archive logs
+      if: always()
+      uses: actions/upload-artifact@v3
+      with:
+        name: multi-axon-nodes-logs
+        path: |
+          ${{ runner.temp }}/${{ matrix.os }}/multi-axon-nodes/
```
Comment (chaoticlonghair) [inline .github/workflows/axon-start-with-short-genesis.yml:121]: I guess there are chain data in this directory.

Due to that, there is an artifact file reached the file size limit in CI.

<details><summary>Error Log Outputs</summary><br>

```
###### End Diagnostic HTTP information ######
Retry limit has been reached for chunk at offset 0 to https://pipelinesghubeus24.actions.githubusercontent.com/kFiqXBRcJNRM3eia9O2QLfAVBdGYeET664SziWqmF7htrjN9jF/_apis/resources/Containers/22326230?itemPath=multi-axon-nodes-logs%2Fnode_3.log
Warning: Aborting upload for /home/runner/work/_temp/ubuntu-20.04/multi-axon-nodes/node_3.log due to failure
Error: aborting artifact upload
Total size of all the files uploaded is 10081 bytes
File upload process has finished. Finalizing the artifact upload
Upload finished. There were 1 items that failed to upload

The raw size of all the files that were specified for upload is 87110 bytes
The size of all the files that were uploaded is 10081 bytes. This takes into account any gzip compression used to reduce the upload size, time and storage

Note: The size of downloaded zips can differ significantly from the reported size. For more information see: https://github.com/actions/upload-artifact#zipped-artifact-downloads 

Error: An error was encountered when uploading multi-axon-nodes-logs. There were 1 items that failed to upload.
```

Ref: https://github.com/axonweb3/axon/actions/runs/5928653600/job/16091034634#step:5:58

</details>

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    - uses: actions/checkout@v3
+
+    - name: Cache of the Axon binary
+      id: axon-bin-cache
+      uses: actions/cache@v3
+      with:
+        path: |
+          target/debug/axon
+          target/release/axon
+        key: ${{ matrix.os }}-${{ runner.os }}-${{ runner.arch }}-axon-${{ github.sha }}
```
Comment (chaoticlonghair) [inline .github/workflows/build.yml:32]: #### Can the cache live longer?

I didn't check the GitHub document, I'm not sure if there is no parameter can to that. If none, then ignore this comment.

#### Why I asked that?

I re-ran the CI this morning, [the cache is missed](https://github.com/axonweb3/axon/actions/runs/5928653600/job/16089183789#step:3:11).
But, [it was created just 9 hours ago](https://github.com/axonweb3/axon/actions/runs/5928653600/job/16089183506#step:9:5
).

File: .github/workflows/axon-start-with-short-genesis.yml (hunk 1 above)
Comment (Flouse) [reply .github/workflows/axon-start-with-short-genesis.yml:121]: > I guess there are chain data in this directory.
> 
> Due to that, there is an artifact file reached the file size limit in CI.

---
> The size of all the files that were uploaded is 10081 bytes.

I don't think so, they are just log files in the `/tmp/...` dir.
Let me re-run it.
Let's see what happens.

File: .github/workflows/build.yml (hunk 2 above)
Comment (Flouse) [reply .github/workflows/build.yml:32]: The reason is https://github.com/axonweb3/axon/actions/workflows/remove_ci_caches.yml
This workflow was disabled by me 5 hours ago.

Please let me know if Axon needs it.

File: .github/workflows/axon-start-with-short-genesis.yml (hunk 1 above)
Comment (Flouse) [reply .github/workflows/axon-start-with-short-genesis.yml:121]: https://github.com/axonweb3/axon/actions/runs/5928653600
test passed.

cc @yangby-cryptape

**Review by chaoticlonghair — APPROVED** [review]

**Review by driftluo — APPROVED** [review]

**Pull request merged**